### PR TITLE
Feature/use from param

### LINKED ISSUE #4427: "Fetching messages..." flashes when there is a contact request in mailbox
### Description

*Type*: Bug

*Summary*: Fetching messages still flashes when you have a contact request in the mailbox. It's because it fetches again the messages for the particular topic that gets added.

#### Expected behavior
Fetching messages status should not be repeated when there is a contact request in chat list. It should not distract a user.

#### Actual behavior
Fetching messages still flashes when you have a contact request in the mailbox.

### Reproduction

- Open Status on 2 devices
- Device 1: Add device 2
- Device 2: Do not accept the contact 

### Additional Information
* Status version: nightly, May 24
* Operating System: Android, iOS

### PR DESCRIPTION
**Pull request by yenda**
fixes #4427
The first commit in this PR can be ignored as it is alreay in #4467 
The second commit introduces a way to differentiate mailserver messages from other whisper messages to keep a timestamp of the last mailserver message received. To do that we compare current time with message time stamp + ttl, if message is expired it comes from the mailserver.
Finally the third commit reimplement a similar mechanism to what was proposed in #4230 except that the :last-request is only updated 10 seconds after the latest message from mailserver has been received without a change of connection state:

When logging in, request messages from last-request for all chats, except chats added while offline for which we request full history. Same when connectivity is re-established after connectivity loss
When adding a chat, request full history for the chat  

# notes for testing
requires some chaos testing
killing the app while fetching, shutting down network while fetching,  joining chat while offline...

status: ready

### CODE REVIEW

File: src/status_im/transport/inbox.cljs [hunk 1]
```diff
@@ -101,7 +108,7 @@
    (mark-trusted-peer web3
                       wnode
                       #(re-frame/dispatch [:inbox/mailserver-trusted %])
-                      #(re-frame/dispatch [:inbox/connection-check]))))
+                      #(re-frame/dispatch [:inbox/check-connexion]))))
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs]: is this a typo?

File: src/status_im/transport/handlers.cljs [hunk 2]
```diff
@@ -18,25 +18,36 @@
             [status-im.transport.message.v1.group-chat :as v1.group-chat]
             [status-im.data-store.transport :as transport-store]))
 
-(defn receive-message [cofx chat-id js-message]
-  (let [{:keys [payload sig]} (js->clj js-message :keywordize-keys true)
+(defn update-last-received-from-inbox
```
Comment (cammellos) [inline src/status_im/transport/handlers.cljs:21]: I haven't quite figured how it works all together, just a rough idea, but just by looking at it:

seems like we should handle out-of-order messages? (`:inbox/last-received` > `timestamp `)

Comment (cammellos) [reply src/status_im/transport/handlers.cljs:21]: Ah no, sorry my bad, you set it to `now`, I thought you were setting it to `timestamp`

File: src/status_im/transport/inbox.cljs [hunk 3]
```diff
@@ -226,20 +288,66 @@
...
+                             (assoc :transport/chats chats-update)
+                             (assoc-in [:account/account :last-request]
+                                       last-request))
+     :data-store/base-tx [(accounts-store/save-account-tx
+                           (assoc (:account/account db)
+                                  :last-request last-request))]
+     :data-store/tx      transport-txs}))
+
+(defn update-fetch-history [chat-id {:keys [db]}]
+  {:db (assoc-in db [:transport/chats chat-id :fetch-history?] false)
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs]: map alignment

File: src/status_im/transport/inbox.cljs [hunk 4]
```diff
@@ -136,69 +157,110 @@
...
+                            :web3       web3}
+        ::request-history {:wnode      wnode
+                           :topics     request-history-topics
+                           :sym-key-id sym-key-id
+                           :web3       web3}
+        :db                (assoc db :inbox/fetching? true)
+        :dispatch-later    [{:ms fetching-timeout
+                             :dispatch [:inbox/check-fetching now-in-s]}]})))
+  ([should-recover? {:keys [db] :as cofx}]
+   (when should-recover?
```
Comment (cammellos) [inline src/status_im/transport/inbox.cljs:267]: Is this  arity being used? I can't find it in the codebase, probably I am missing something

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/voN3bH for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/voN3bH">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

File: src/status_im/transport/inbox.cljs (hunk 4 above)
Comment (yenda) [reply src/status_im/transport/inbox.cljs:267]: here https://github.com/status-im/status-react/pull/4481/files#diff-9c6dc8fc2461b89e7486af5f1fe7d2f3R459

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/1d5jaH for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/1d5jaH">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

File: src/status_im/transport/inbox.cljs [hunk 5]
```diff
@@ -136,69 +157,110 @@
...
+  [db]
+  (conj (mapv :topic
+              (remove :fetch-history?
+                      (vals (:transport/chats db))))
+        (transport.utils/get-topic constants/contact-discovery)))
+
+(defn get-request-history-topics
+  "Returns topics for which full history has not been recovered"
+  [db]
+  (mapv :topic
```
Comment (dmitryn) [inline src/status_im/transport/inbox.cljs:238]: is there a reason to use `mapv`, but not `map` here? Code is similar to the above in `get-request-messages-topics` fn where it really needs vector to `conj`  with

File: src/status_im/transport/inbox.cljs (hunk 4 above)
Comment (cammellos) [reply src/status_im/transport/inbox.cljs:267]: thanks!

**Review by janherich — APPROVED** [review]

**Review by dmitryn — APPROVED** [review]

File: src/status_im/transport/inbox.cljs [hunk 6]
```diff
@@ -115,9 +122,23 @@
...
+(re-frame/reg-fx
+ ::request-history
+ (fn [{:keys [wnode topics to from sym-key-id web3]}]
+   (request-inbox-messages web3
+                           wnode
+                           topics
+                           to
+                           from
+                           sym-key-id
+                           #(log/info "offline inbox: request-messages response" %)
```
Comment (dmitryn) [inline src/status_im/transport/inbox.cljs:152]: Probably not part of this PR, but current log information print `topics` being requested, but not `from` / `to` parameters. Would be helpful to log all the parameters being used for request.

Comment (yenda) [reply src/status_im/transport/inbox.cljs:152]: I will take care of this when doing https://github.com/status-im/status-react/issues/4387

File: src/status_im/transport/inbox.cljs (hunk 5 above)
Comment (yenda) [reply src/status_im/transport/inbox.cljs:238]: it's just for consistency, lazyness doesn't matter here anyway

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/1TtmyX for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/1TtmyX">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/bgD8V5 for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/bgD8V5">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Review by cammellos — APPROVED** [review]
Overall looks ok, 

We should probably document the way the algorithm works, either through ADR/integration tests or wiki, as it is not immediately obvious from the code (that's just by nature of the events/asynchronous nature of it, rather than the implementation).

I have one question, why we chose to use the time we receive the last message to increment the pagination counter, as opposed to save it to the db after we fire the request?

This introduces a few cases we need to watch out for / some complexity, such as correctly detecting expired messages (we might want to leave some headroom as it might be that status react process a message which has expired locally), and network latency might skew pagination, creating gaps. 

Maybe this kind of decision are good to document in an ADR explaining why you chose the algorithm, what it solves and what it does not so that we are also aware when making changes.

What do you think?

**Comment by cammellos** [comment]
I stand corrected, I mistakenly read the code and assumed `last-received` was used to fetch as well, ignore the comment above regarding the 2 implementations, sorry for the confusion!

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/DM2QTh for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/DM2QTh">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/93qQPu for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/93qQPu">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by yenda** [comment]
I added a comment explaining how it works at the beginning of the namespace to clarify how it works

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/YZXvcw for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/YZXvcw">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/pR9DM9 for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/pR9DM9">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by lukaszfryc** [comment]
@yenda I tested all kind of weird scenarios (offline -> online) related to chats including contacts requests and receiving message notification when app was in the background and I went offline -> online. All of them worked great 👍 

But, when I go offline, send contact request and then go online it's not received on the second device. It maybe that it is not being sent after going online. Could you check it?

**Comment by yenda** [comment]
@lukaszfryc is the contact request thing the only issue ? It is not due to the PR https://github.com/status-im/status-react/issues/4212 , the contact-request is never sent if you create the contact while you are offline so that is not something that is in the scope of this PR, but I will work on it after this.

**Comment by lukaszfryc** [comment]
> the contact-request is never sent if you create the contact while you are offline so that is not something that is in the scope of this PR, but I will work on it after this.


@yenda when https://github.com/status-im/status-react/pull/4336 is merged, the contact requests won't be shown on receiver's device. So, maybe it's not worth fixing it? Will the fix have any positive impact on user experience?

**Comment by rasom** [comment]
apk uploaded to https://i.diawi.com/5ZYciX for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/5ZYciX">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>

**Comment by rasom** [comment]
ipa uploaded to https://i.diawi.com/kEKH8S for branch PR-4481

<blockquote><div><strong><a href="https://i.diawi.com/kEKH8S">Install - Diawi - Development and In-house Apps Wireless Installation</a></strong></div></blockquote>
